In [1]:
import pandas as pd

In [2]:
purchases = pd.read_csv("../Resources/purchase_data.csv")
purchases.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

First off, we're going to figure out how many unique players have visited this online shop

In [3]:
#Used to count the amount of players
player_count = 0
#Used to store the names of unique players
unique_count = []
#and their gender, for later use
gender = []
#and their age, also for later use
age = []
#loop through SN column for unique user ID's
for i in range(0, purchases["SN"].count()):
    if purchases["SN"][i] not in unique_count:
        #add user ID's, the gender, and age to their respective lists
        unique_count.append(purchases["SN"][i])
        gender.append(purchases["Gender"][i])
        age.append(purchases["Age"][i])
        #increment the player count
        player_count +=1
print(player_count)

576


In [41]:
#A dataframe containing the unique players
new = {"Names": unique_count, "Gender": gender, "Age": age}
newdf = pd.DataFrame(new)
newdf

Names  Gender  Age
0          Lisim78    Male   20
1      Lisovynya38    Male   40
2       Ithergue48    Male   24
3    Chamassasya86    Male   24
4        Iskosia90    Male   23
..             ...     ...  ...
571         Hala31    Male   21
572     Jiskjask80    Male   11
573     Aethedru70  Female   21
574     Yathecal72    Male   20
575        Sisur91    Male    7

[576 rows x 3 columns]

There seem to be a total of 576 unique customer who use the shop.
## Purchasing Analysis (Total)
Now let's clean up the data and figure out:
* The number of unique items in the shop
* The average price of items
* The total number of purchases
* Total revenue

In [4]:
#initiate variables for counting unique items
items = 0
#Because i'm uncreative with names
item_list = []
#loop through the Item Name column for unique item names
for i in range(0, purchases["Item Name"].count()):
    if purchases["Item Name"][i] not in item_list:
        #add item names to the  list, which also increments the items count
        item_list.append(purchases["Item Name"][i])
        items += 1
print(f"We sell {items} items in our shop!")

We sell 179 items in our shop!


In [5]:
#Total Revenue
price_total = purchases["Price"].sum()
print(f"We made ${price_total} today!")

We made $2379.77 today!


In [6]:
#Number of purchases
purch = purchases["Purchase ID"].count()
print(f"We had {purch} visits today!")

We had 780 visits today!


In [7]:
#Average Price
avg_price = round((price_total / purch), 2)
print(f"The average price is ${avg_price}.")

The average price is $3.05.


In [8]:
#Put it all in a neat DataFrame
purchase_analysis = {"Number of Unique Items": [items],
                     "Purchases": [purch],
                     "Total Revenue": [price_total],
                     "Average Price": [avg_price]}
df = pd.DataFrame(purchase_analysis)
print(df)

   Number of Unique Items  Purchases  Total Revenue  Average Price
0                     179        780        2379.77           3.05


## Gender Demographics

Now that we got our purchase analysis, next up is checking out the gender demographic.  

In [9]:
#create a dataframe that takes each unique name as a key and the gender as a value
dicts = {"Name": unique_count, "Gender": gender}
df = pd.DataFrame(dicts)
df

Name  Gender
0          Lisim78    Male
1      Lisovynya38    Male
2       Ithergue48    Male
3    Chamassasya86    Male
4        Iskosia90    Male
..             ...     ...
571         Hala31    Male
572     Jiskjask80    Male
573     Aethedru70  Female
574     Yathecal72    Male
575        Sisur91    Male

[576 rows x 2 columns]

In [10]:
#initiate variables
male = 0
female = 0
other = 0
#initiate unimaginative lists
male_list = []
female_list = []
other_list = []
#loop through the Gender column
for i in range(0, df["Gender"].count()):
    if df["Gender"][i] == "Male":#increment the male count
        male_list.append(purchases["Gender"][i])
        male += 1
    elif df["Gender"][i] == "Female":#increment the female count
        female_list.append(purchases["Gender"][i])
        female += 1
    else:#increment the other count
        other_list.append(purchases["Gender"][i])
        other += 1
print(f"There are {male} males, {female} females, and {other} of non-disclosed gender.")


There are 484 males, 81 females, and 11 of non-disclosed gender.


In [11]:
#determine the ratios of each gender 
male_ratio = round((male / player_count * 100), 2)
female_ratio = round((female / player_count * 100), 2)
other_ratio = round((other / player_count * 100), 2)
print(f"The ratios are as follows: {male_ratio}% males, {female_ratio}% females, and {other_ratio}% of undisclosed gender.")

The ratios are as follows: 84.03% males, 14.06% females, and 1.91% of undisclosed gender.


In [12]:
#clean it up as a DataFrame
gender_data = {"Total Count": [male, female, other], 
               "Percentage of Players": [male_ratio, female_ratio, other_ratio]}
gender_df = pd.DataFrame(gender_data, index = ["Male", "Female", "Other / Undisclosed"])
gender_df

Total Count  Percentage of Players
Male                         484                  84.03
Female                        81                  14.06
Other / Undisclosed           11                   1.91

## Purchasing Analysis (Gender)
We've broken down the amount of unique players by gender as well as the ratio of each gender.
Now we're going to analyze the average purchases made by each gender

In [13]:
#reinitiate variables
remale = 0
refemale = 0
reother = 0
#reinitiate unimaginative lists
remale_list = []
refemale_list = []
reother_list = []
#a surprise tool to help us later
male_price = []
female_price = []
other_price = []
#reloop through the Gender column
for i in range(0, purchases["Gender"].count()):
    if purchases["Gender"][i] == "Male":#increment the total male count
        remale_list.append(purchases["Gender"][i])
        male_price.append(purchases["Price"][i])
        remale += 1
    elif purchases["Gender"][i] == "Female":#increment the total female count
        refemale_list.append(purchases["Gender"][i])
        female_price.append(purchases["Price"][i])
        refemale += 1
    else:#increment the total other count
        reother_list.append(purchases["Gender"][i])
        other_price.append(purchases["Price"][i])
        reother += 1
print(f"There are {remale} males, {refemale} females, and {reother} of non-disclosed gender.")


There are 652 males, 113 females, and 15 of non-disclosed gender.


In [14]:
#total purchasing value per gender
male_total_purchase = round(sum(male_price), 2)
female_total_purchase = round(sum(female_price), 2)
other_total_purchase = round(sum(other_price), 2)
print(f"Totals with males having ${male_total_purchase}, females having ${female_total_purchase}, and undisclosed genders with ${other_total_purchase}.")

Totals with males having $1967.64, females having $361.94, and undisclosed genders with $50.19.


In [15]:
#figuring out the average purchasing price per gender
avg_male_purch = round((male_total_purchase / remale), 2)
avg_female_purch = round((female_total_purchase / refemale), 2)
avg_other_purch = round((other_total_purchase / reother), 2)
print(f"An average purchase price of ${avg_male_purch} for males, ${avg_female_purch} for females, and ${avg_other_purch} for undisclosed gender.")

An average purchase price of $3.02 for males, $3.2 for females, and $3.35 for undisclosed gender.


In [16]:
#figuring out the average total purchase per person
avg_total_per_male = round((male_total_purchase / male), 2)
avg_total_per_female = round((female_total_purchase / female), 2)
avg_total_per_other = round((other_total_purchase / other), 2)
print(f"An average total purchase per person of ${avg_total_per_male} for males, ${avg_total_per_female} for females, and ${avg_total_per_other} for undisclosed genders.")

An average total purchase per person of $4.07 for males, $4.47 for females, and $4.56 for undisclosed genders.


In [17]:
#Let's clean it up with a DataFrame
gender_purchase_analysis = {#"Gender": ["Male", "Female", "Other / Undisclosed"],
                            "Purchase Count": [remale, refemale, reother],
                            "Total Purchase Value in Dollars": [male_total_purchase, female_total_purchase, other_total_purchase],
                            "Average Purchase Price in Dollars": [avg_male_purch, avg_female_purch, avg_other_purch],
                            "Average Total Purchase per Person in Dollars": [avg_total_per_male, avg_total_per_female, avg_total_per_other]
                            }
gender_purch_df = pd.DataFrame(gender_purchase_analysis, index = ["Male", "Female", "Other / Undisclosed"])
index = gender_purch_df.index
index.name = "Gender"
gender_purch_df

Purchase Count  Total Purchase Value in Dollars  \
Gender                                                                 
Male                            652                          1967.64   
Female                          113                           361.94   
Other / Undisclosed              15                            50.19   

                     Average Purchase Price in Dollars  \
Gender                                                   
Male                                              3.02   
Female                                            3.20   
Other / Undisclosed                               3.35   

                     Average Total Purchase per Person in Dollars  
Gender                                                             
Male                                                         4.07  
Female                                                       4.47  
Other / Undisclosed                                          4.56

## Age Demographics
Of course, players of many ages are playing Heroes of Pymoli.  So we want to determine the ages of our players, and how many of them take up the playerspace. 

In [39]:
#Separate the Age column and bin certain ages 
binned = [0, 9, 14, 19, 24, 29, 34, 39, 100]
#labels for the bins
lab = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]
#cut up the column and create a new "binned" column for the ages
bigbin = pd.cut(newdf["Age"], bins=binned, labels=lab)
bigbin

0      20-24
1        40+
2      20-24
3      20-24
4      20-24
       ...  
571    20-24
572    10-14
573    20-24
574    20-24
575      <10
Name: Age, Length: 576, dtype: category
Categories (8, object): [<10 < 10-14 < 15-19 < 20-24 < 25-29 < 30-34 < 35-39 < 40+]

In [43]:
#simplify the binned column
ages = bigbin
#it is variable time
kid = 0
preteen = 0
teen = 0
young_adult = 0
adult_iguess = 0
adult_2 = 0
adult_3 = 0
old = 0
#loop through the binned column
for i in range(0, ages.count()):
    #for every bin discovered, increment its respective count
    if ages[i] == "<10":
        kid += 1
    elif ages[i] == "10-14":
        preteen += 1
    elif ages[i] == "15-19":
        teen += 1
    elif ages[i] == "20-24":
        young_adult += 1
    elif ages[i] == "25-29":
        adult_iguess += 1
    elif ages[i] == "30-34":
        adult_2 += 1
    elif ages[i] == "35-39":
        adult_3 += 1
    else:
        old += 1
print(f"We've got {kid} under 10, {preteen} people 10-14, {teen} people 15-19, {young_adult} people 20-24, {adult_iguess} people 25-29, {adult_2} people 30-34, {adult_3} people 35-39, and {old} people 40+.")

We've got 17 under 10, 22 people 10-14, 107 people 15-19, 258 people 20-24, 77 people 25-29, 52 people 30-34, 31 people 35-39, and 12 people 40+.


In [50]:
#time for percentages; calculate the percentage of each age category
kid_perc = round((kid / player_count * 100), 2)
preteen_perc = round((preteen / player_count * 100), 2)
teen_perc = round((teen / player_count * 100), 2)
young_adult_perc = round((young_adult / player_count * 100), 2)
adult_iguess_perc = round((adult_iguess / player_count * 100), 2)
adult_2_perc = round((adult_2 / player_count * 100), 2)
adult_3_perc = round((adult_3 / player_count * 100), 2)
old_perc = round((old / player_count * 100), 2)

In [52]:
#And a DataFrame to clean it up
age_demo = {"Total Count": [kid, preteen, teen, young_adult, adult_iguess, adult_2, adult_3, old],
            "Percentage of Players": [kid_perc, preteen_perc, teen_perc, young_adult_perc, adult_iguess_perc, adult_2_perc, adult_3_perc, old_perc]
           }
age_demo_df = pd.DataFrame(age_demo, index = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"])
age_demo_df

Total Count  Percentage of Players
<10             17                   2.95
10-14           22                   3.82
15-19          107                  18.58
20-24          258                  44.79
25-29           77                  13.37
30-34           52                   9.03
35-39           31                   5.38
40+             12                   2.08